lib/Runtime/Language/JavascriptOperators.cpp

```diff
@@ -4418,45 +4418,39 @@ namespace Js
 
         if (indexType == IndexType_Number)
         {
+SetElementIHelper_INDEX_TYPE_IS_NUMBER:
             return JavascriptOperators::SetItem(receiver, object, indexVal, value, scriptContext, flags);
         }
         else if (indexType == IndexType_JavascriptString)
         {
             Assert(propertyNameString);
-            JsUtil::CharacterBuffer<WCHAR> propertyName(propertyNameString->GetString(), propertyNameString->GetLength());
 
-            if (BuiltInPropertyRecords::NaN.Equals(propertyName))
-            {
-                // Follow SetProperty convention for NaN
-                return JavascriptOperators::SetProperty(receiver, object, PropertyIds::NaN, value, scriptContext, flags);
-            }
-            else if (BuiltInPropertyRecords::Infinity.Equals(propertyName))
-            {
-                // Follow SetProperty convention for Infinity
-                return JavascriptOperators::SetProperty(receiver, object, PropertyIds::Infinity, value, scriptContext, flags);
-            }
-#ifdef ENABLE_DEBUG_CONFIG_OPTIONS
-            if (PHASE_TRACE1(PropertyStringCachePhase))
+            // At this point, we know that the propertyNameString is neither PropertyString
+            // or LiteralStringWithPropertyStringPtr.. Get PropertyRecord!
+            // we will get it anyways otherwise. (Also, 1:1 string comparison for Builtin types will be expensive.)
+
+            if (propertyRecord == nullptr)
             {
-                Output::Print(_u("PropertyCache: SetElem No property string for '%s'\n"), propertyNameString->GetString());
+                scriptContext->GetOrAddPropertyRecord(propertyNameString, &propertyRecord);
+                if (propertyRecord->IsNumeric())
+                {
+                    indexVal = propertyRecord->GetNumericValue();
+                    goto SetElementIHelper_INDEX_TYPE_IS_NUMBER;
```

Comment:
I feel like the if conditions could be structured here such that we wouldn't need a backwards goto.

Reply:
Since it is just going to a return statement, I think it would be much clearer to just put the statement here and let the compiler combine them together if it wanted to.

Reply:
duplicate the code? why it is more clear?

Reply:
I tend to feel that gotos add complexity/anxiety to a codebase; there are definitely times when they are useful, but I don't see the benefit here comparing `goto SetElementIHelper_INDEX_TYPE_IS_NUMBER;` and `return JavascriptOperators::SetItem(receiver, object, indexVal, value, scriptContext, flags);`. With the return it's obvious what is happening, while with the goto I need to go look at another piece of code, and potentially try and understand if there are variables that change scopes and have unexpected behavior.

Reply:
well, I'm not strong on either way. I just don't like copy-paste code all over the place.

Anything else?

Reply:
* I don't have strong opinion one way or another..